tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,276 @@
+  return visited_.contains(instruction);
+}
+
+absl::optional<std::vector<HloInstruction*>>
+ SliceSinkerImpl::FindSourceOperandsOfSlicesForElementwiseOperation(
+    const HloInstruction* inst) {
+  std::vector<HloInstruction*> operands;
+  // Check operand:
+  // The inst's operand should be a slice of the true operand.
+  // The operands-vector keeps all true operands of inst.
+  for (HloInstruction* operand_slice : inst->operands()) {
+    if (operand_slice->opcode() != HloOpcode::kSlice) {
+      // If the operand is not operand-slice, returns empty vector.
+      visited_.insert(inst);
+      return absl::nullopt;
+    }
+    // The operand-slice's operand is a true operand.
+    HloInstruction* operand = operand_slice->mutable_operand(0);
+    operands.push_back(operand);
+  }
+
+  // No true operands found. returns empty vector.
+  if (operands.empty()) {
+    visited_.insert(inst);
+    return absl::nullopt;
+  }
+
+  // Check operands:
+  // True operands should have the same shape because inst is elementwise.
+  const Shape shape = operands[0]->shape();
+  for (const HloInstruction* operand : operands) {
+    // Only support element-wise now
+    if (!ShapeUtil::CompatibleIgnoringElementType(operand->shape(), shape)) {
+      visited_.insert(inst);
+      return absl::nullopt;
+    }
+  }
+  // Inst's operand-slices should be sliced from the same location of true
+  // operands.
+  const HloInstruction* operand0 = inst->operand(0);
+  for (const HloInstruction* operand : inst->operands()) {
+    if (operand0->slice_starts() != operand->slice_starts() ||
+        operand0->slice_limits() != operand->slice_limits() ||
+        operand0->slice_strides() != operand->slice_strides()) {
+      visited_.insert(inst);
+      return absl::nullopt;
+    }
+  }
+  // Returns all true operands of inst.
+  return operands;
+}
+
+absl::optional<std::vector<HloInstruction*>>
+SliceSinkerImpl::FindPeerElementwiseOperations(
+    const HloInstruction* inst,
+    const std::vector<HloInstruction*>& operands) {
+  std::vector<HloInstruction*> users;
+  HloInstruction* operand0 = operands[0];
+
+  // Traverses a true operand's all operand-slices.
+  for (const HloInstruction* operand_slice0 : operand0->users()) {
+    // Skips not-slices
+    if (operand_slice0->opcode() != HloOpcode::kSlice) {
+      continue;
+    }
+
+    // The user of operand-slices is a candidate of true user.
+    for (HloInstruction* user : operand_slice0->users()) {
+      // The inst's peers should be the same operation and same operand count as
+      // inst. The visited user is in another group of other true users. Skips
+      // them to avoid useless match.
+      if (IsVisited(user) || user->opcode() != inst->opcode() ||
+          user->operand_count() != inst->operand_count()) {
+        continue;
+      }
+      // Checks operand-slices:
+      bool isTrueUser = true;
+      for (int64 j = 0; j < operands.size(); ++j) {
```

Comment:
Can we outline this loop to a routine like this:
bool IsPeerInstruction(const HloInstruction* peer_candidate, HloOpcode opcode, absl::Span<HloInstruction*> slice_sources)
There is no need to pass operand_slice0 to the routine, even though it is used in this loop, because operand_slice0 is just an operand of peer_candidate.